server/athenian/api/experiments/preloading/entries.py

```diff
@@ -1,8 +1,137 @@
-from athenian.api.controllers.features.entries import \
-    MetricEntriesCalculator as OriginalMetricEntriesCalculator
+from datetime import datetime
+from typing import Collection, Optional, Set, Tuple
+
+import aiomcache
+import databases
+import pandas as pd
+
+from athenian.api.controllers.features.entries import (
+    MetricEntriesCalculator as OriginalMetricEntriesCalculator,
+)
+from athenian.api.controllers.miners.filters import JIRAFilter, LabelFilter
+from athenian.api.controllers.miners.github.pull_request import (
+    PullRequestMiner as OriginalPullRequestMiner,
+)
+from athenian.api.controllers.miners.types import PRParticipants, PRParticipationKind
+from athenian.api.models.metadata.github import PullRequest
+from athenian.api.tracing import sentry_span
+
+
+class PullRequestMiner(OriginalPullRequestMiner):
+    """Load all the information related to Pull Requests from the metadata DB. Iterate over it \
+    to access individual PR objects."""
+
+    @classmethod
+    @sentry_span
+    async def _fetch_prs_by_filters(
+        cls,
+        time_from: datetime,
+        time_to: datetime,
+        repositories: Set[str],
+        participants: PRParticipants,
+        labels: LabelFilter,
+        jira: JIRAFilter,
+        exclude_inactive: bool,
+        pr_blacklist: Optional[Collection[str]],
+        pr_whitelist: Optional[Collection[str]],
+        meta_ids: Tuple[int, ...],
+        mdb: databases.Database,
+        cache: Optional[aiomcache.Client],
+        columns=PullRequest,
+        updated_min: Optional[datetime] = None,
+        updated_max: Optional[datetime] = None,
+    ) -> pd.DataFrame:
+        if jira or labels:
+            # TODO: not supported yet, call original implementation
+            return await super(PullRequestMiner, cls)._fetch_prs_by_filters(
+                time_from,
+                time_to,
+                repositories,
+                participants,
+                labels,
+                jira,
+                exclude_inactive,
+                pr_blacklist,
+                pr_whitelist,
+                meta_ids,
+                mdb,
+                cache,
+                columns=columns,
+                updated_min=updated_min,
+                updated_max=updated_max,
+            )
+
+        assert (updated_min is None) == (updated_max is None)
+
+        df = mdb.cache.dfs["prs"].df
+
+        mask = (
+            (~df["closed"] | (df["closed_at"] >= time_from))
+            & (df["created_at"] < time_to)
+            & (df["acc_id"].isin(meta_ids))
+            & (~df["hidden"])
+            & (df["repository_full_name"].isin(repositories))
+        )
+
+        if pr_blacklist is not None:
```

Comment:
I see why you had to switch from BinaryExpression to plain lists. I can help you with traversing the sqla AST here, I did it a few times already. It's not so difficult actually.

Reply:
What about this?
```
In [29]: whitelist = PullRequest.node_id.in_(['abc','def'])

In [30]: blacklist = PullRequest.node_id.notin_(['ghi','jkl'])

In [31]: [bp.value for bp in whitelist.get_children()[1].get_children()[0].get_children()]
Out[31]: ['abc', 'def']

In [32]: [bp.value for bp in blacklist.get_children()[1].get_children()[0].get_children()]
Out[32]: ['ghi', 'jkl']
```

Reply:
Yes, smth like this, exactly. They can be `in_values` and `not_in_values`, too, we should check that they work the same. Those are defined in `athenian/api/models/__init__.py`

Reply:
```
In [4]: extract = lambda be: [bp.value for bp in be.get_children()[1].get_children()[0].get_children()]

In [5]: extract(PullRequest.node_id.in_(['abc','def']))
Out[5]: ['abc', 'def']

In [6]: extract(PullRequest.node_id.notin_(['abc','def']))
Out[6]: ['abc', 'def']

In [7]: extract(PullRequest.node_id.in_any_values(['abc','def']))
Out[7]: ['abc', 'def']

In [8]: extract(PullRequest.node_id.notin_any_values(['abc','def']))
Out[8]: ['abc', 'def']
```

confirmed 👍